src/eins-hwinfo.c

```diff
@@ -0,0 +1,119 @@
+ *
+ * eos-metrics-instrumentation is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-metrics-instrumentation.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+#include "eins-hwinfo.h"
+
+#include <eosmetrics/eosmetrics.h>
+#include <gio/gio.h>
+
+/*
+ * Reported at system startup, and every RECORD_DISK_SPACE_INTERVAL seconds
+ * after startup. The payload is two triples of int64, ((ttt)(ttt)).
+ *
+ * The first triple represents the total size, space used, and space available
+ * on the / filesystem, measured in bytes. (On dual-boot installations, this
+ * refers to the Endless OS image file, not the Windows partition it is hosted
+ * on.)
+ *
+ * On split-disk systems, the second triple represents the /var/endless-extra
+ * partition; on single-disk systems, all three values will be 0.
+ *
+ * Space on other user-accessible partitions on the disk, including Windows
+ * partitions on dual-boot systems, is not reported.
+ *
+ * You might think that given any two of these values for a filesystem, you
+ * could derive the third. That's not the case: typically, 5% of space is
+ * reserved (so used + available = 0.95 * total) but this is a tunable
+ * parameter of the filesystem.
+ */
+#define DISK_SPACE_EVENT "b76c645f-d041-41b0-a4d4-c29f86c18638"
+
+#define RECORD_DISK_SPACE_INTERVAL (60u * 60u * 24u)
+
+static GVariant *
+get_disk_space_for_partition (GFile *file)
+{
+  g_autoptr(GFileInfo) info = NULL;
+  g_autoptr(GError) error = NULL;
+  guint64 total = 0;
+  guint64 used = 0;
+  guint64 available = 0;
+
+  info = g_file_query_filesystem_info (file,
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_SIZE ","
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_USED ","
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_FREE,
+                                       NULL /* cancellable */,
+                                       &error);
+  if (info == NULL)
+    {
+      g_warning ("%s: %s", G_STRFUNC, error->message);
+    }
+  else
+    {
+      total = g_file_info_get_attribute_uint64 (info,
+                                                G_FILE_ATTRIBUTE_FILESYSTEM_SIZE);
+      used = g_file_info_get_attribute_uint64 (info,
+                                               G_FILE_ATTRIBUTE_FILESYSTEM_USED);
+      available = g_file_info_get_attribute_uint64 (info,
+                                                    G_FILE_ATTRIBUTE_FILESYSTEM_FREE);
+    }
+
+  return g_variant_new ("(ttt)", total, used, available);
+}
+
+static gboolean
+is_mounted (GFile *file)
+{
+  g_autoptr(GFileInfo) info = g_file_query_info (file,
+                                                 G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT,
```

Comment:
I wonder if this is really the best way to perform this check; `g_file_find_enclosing_mount()` may be more desirable, since I don't know if this would return positively when e.g. the path is an unmounted mountpoint mentioned in `/etc/fstab`

Reply:
`G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT` is true if `file`'s parent directory is on a different `st_dev` to `file`. (It's currently `FALSE` for `/` which I think [is a bug](https://gitlab.gnome.org/GNOME/glib/issues/1400).)

`g_file_find_enclosing_mount()` just performs a recursive version of the same check, walking all ancestors until it finds a `st_dev` transition or the root directory. Neither one checks `/etc/fstab`.